components/salsa-macro-rules/src/setup_tracked_fn.rs

```diff
@@ -179,12 +182,16 @@ macro_rules! setup_tracked_fn {
                     $inner($db, $($input_id),*)
                 }
 
+                fn cycle_initial<$db_lt>(db: &$db_lt dyn $Db) -> Self::Output<$db_lt> {
+                    $($cycle_recovery_initial)*(db)
```

Comment:
Hmm. In the use cases I have in mind (e.g. for red-knot) there should not be any reason to call a query from within one of these functions. And it seems like this could be quite difficult to deal with. So unless we have clear use cases, I would rather consider this out of scope. We could just document "don't do that," or we could add some kind of explicit prevention.

Reply:
Yeah. I'm not suggesting that users should do that. I only wonder about what happens if a user does it regardless. 

Ideally, we wouldn't provide a `Db` but we can't do that because a user might want to create a tracked struct.

Reply:
I think the simplest approach here would be to add some state on `Runtime` that causes an error if you try to push a new active query. I'll wait for Niko review before doing that, though.

Reply:
Added a TODO comment on this to get Niko's feedback.

Reply:
If this replaces the old cycle handling, we (rust-analyzer) actually needs to be able to call queries here as we already do that in some of our cycle recovery functions today. I think most of those are in part because of assumptions else where in the code, an example being `generic_defaults` always returning a list with length matching the number of generics, and to compute the number we need to call a query. That query can run into a cycle though, so we still need to figure out the number of generics in the recovery function which means we need to call the corresponding query.